### ISSUE TITLE
[Overview] Modernization code with clang-tidy

### ISSUE THREAD

**Issue by SunBlack** [PointCloudLibrary/pcl#2731]
This issue is just to have an overview about clang-tidy modernizations.

Full list of [clang-tidy checks](https://clang.llvm.org/extra/clang-tidy/checks/list.html).

Will be applied:
* [ ] modernize-avoid-c-arrays
* [ ] modernize-deprecated-headers
* [x] modernize-use-override (PR: #2728)

Needs discussion:
* modernize-use-auto

Will not be applied:
* modernize-concat-nested-namespaces (requires C++17)

In case you see a modernization we could apply or can't apply, feel free to write it as comment. In case there is a modernization that needs discussion, open a new issue and write reference to it as comment, so I can add it list.

**Comment by jasjuang** [comment]
I think modernize-use-auto makes a lot of sense for iterators, but for new expressions and cast expressions to make sense, MinTypeNameLength has to be at least larger or equal the default 5. 

modernize-avoid-c-arrays and modernize-deprecated-headers is definitely needed.

I think modernize-make-shared, modernize-make-unique, modernize-use-emplace, modernize-use-nullptr are also good to add in.

**Comment by SunBlack** [comment]
@jasjuang modernize-make-shared & modernize-make-unique will not have any effect, because we didn't had C++11 before ;).

**Comment by SunBlack** [comment]
I added two points for discussion.

**Comment by SunBlack** [comment]
> I think modernize-use-auto makes a lot of sense for iterators, but for new expressions and cast expressions to make sense, MinTypeNameLength has to be at least larger or equal the default 5.

@jasjuang Discussion is for this is currently here: #2838 - in case you want part of it ;)

**Comment by denix56** [comment]
Maybe we should also replace `typedef` with `using` ? Clang also suggests it as modernization. Moreover, it is, imho, much more readable

**Comment by taketwo** [comment]
It'd certainly be nice to switch to `using` in the long run. @SunBlack already attempted to apply automatic fix, but there were problems, see #2791.

**Comment by SunBlack** [comment]
I suggest to skip `modernize-use-default-member-init`, because I really don't like to have default values stored in header.

**Comment by taketwo** [comment]
Generally, I agree with you on this. However, due to the fact that PCL is a heavily templated library, we already have default values in ".h" files most of the times. And even if we don't have them in ".h", then they are in ".hpp", which is effectively the same. Meantime, there are benefits in using default member init. For example, when there are multiple constructors (which is often the case in PCL) it helps to maintain consistency between default values.

**Comment by SergioRAgostinho** [comment]
Just wanted to add here the decision we took for `readability-implicit-bool-conversion`. tldr activate the option `AllowPointerConditions` at that time. See this [comment](https://github.com/PointCloudLibrary/pcl/pull/3029#discussion_r278668564) for reference

**Comment by SunBlack** [comment]
Just to mention: I don't believe we can integrate clang-tidy in near feature into build server. I tried it in our project, but there are some issues:
- I have no idea how `header-filter` really works. It should be a Regex, but I don't know which syntax they allow, because all my regexe works just partially (need to exclude `moc` and `ui` files of Qt from output)
- If we switch to `AUTOMOC`, ÀUTOUI` and ÀUTORCC`, these file will not anymore generated during CMake generation phase. Instead they will be generated at start of a build (there are `_autogen` targets then). So we need to build all `_autogen` targets to get run clang-tidy (if not clang-tidy will complain missing moc/ui/rcc files). Because there is no global `autogen` target, we need to write a script to grep all `_autogen` targets of make file, because makefiles doesn't support wildcard, so we can't call `make *_autogen`. Easiest solution use CMake 3.14 and `CMAKE_EXPORT_COMPILE_COMMANDS=ON`
- Output of clang-tidy 6.0 / 7 / 8 produces sometimes corrupted output, if you parse console output (starts printing output a hit within another hit). So in case we want to parse output, we need to use a YAML parser (this output seems to be valid)
- Not sure how long a run will need on Azure, could be longer than MSVC build ;).

**Comment by taketwo** [comment]
> Not sure how long a run will need on Azure, could be longer than MSVC build ;).

:scream: In case it turns out to be very expensive to run, we may consider making it a periodic job which runs, say, weekly.

But the rest of the items you posted make this all sound like a very daunting task.

**Comment by SunBlack** [comment]
General question, due to:

> Lack of time to review. Like I said, I'm not putting any priority in reviewing these PRs because they are time consuming and not in the goals for this release. That being said, they are fairly trivial in the changes they make, so feel I'm perfectly ok in merging them with just your review.

_Originally posted by @SergioRAgostinho in https://github.com/PointCloudLibrary/pcl/pull/3112#issuecomment-500226328_

Should I stop applying modernization until release of 1.10.0 or still continue?

**Comment by SergioRAgostinho** [comment]
If you have time to spare, I would rather have your help on the transition to boost until the release is complete. There are a couple of items now that are simply manual labor, e.g.: the bind -> lambda replacement. But let Sergey express his opinion, since he was still reviewing things.

**Comment by taketwo** [comment]
I am fine reviewing/merging clang-tidy conversions. But indeed help on the milestone items would be more important.

**Comment by SunBlack** [comment]
Well ok, don't know if I really have (currently) time to help on boost. There is a small thing I will change, if I have time (PCL is using boost filesystem still via iterator instead of for-ranged loop).

Nevertheless I will continue with some clang-tidy changes, because I can run clang-tidy if I don't need my PC and just need to review it after it .

**Comment by SunBlack** [comment]
Updated list, so progress is better to see.

**Comment by SunBlack** [comment]
I did run this night clang-tidy with all checks on our build server, to get a better feeling, which checks will produce smaller PRs and which bigger one. Will, this did run a bit out of control 😆 

Job statistic:
* Error Log-File Size of clang-tidy: 287.71MB
* clang-tidy time: 1h 13min
* Java parsing time: Still noch finished after > 7h

General warnings statistic:
* Total: 1170100
* Outside of PCL code: 237696
* Within the 3rd-Party code of PCL: 67130
* Within the non-3rd-Party code of PCL: 865274

Categories:
* abseil: 0
* android: 2
* boost: 0
* bugprone: 6461
* cert: 2834
* clang-analyzer: 791
* cppcoreguidelines: 246098
* fuchsia: 153825
* google: 151613
* hicpp: 82940
* llvm: 46737
* misc: 37543
* modernize: 35527
* mpi: 0
* objc: 0
* performance: 1367
* portability: 2690
* readability: 100747
* zircon: 0

**Comment by SergioRAgostinho** [comment]
Sheer curiosity: what exactly is your company doing with PCL? I'm curious on the reasons for subjecting all your code base to this extensive battery of checks. 👀

**Comment by SunBlack** [comment]
> Sheer curiosity: what exactly is your company doing with PCL?

We are not using this much modules of PCL. Most reason to use PCL for us: If we want to experiment with data we have here already a lot of algorithms, so we don't need to implement everything on our own.

>I'm curious on the reasons for subjecting all your code base to this extensive battery of checks. 👀

We are working with a lot of students, so we have a lot of developer who only spend a few month/years on our project. So automatic checks are perfect to reduce time we maintainer have to spend on MRs to get the code ready to merge. In general: My main task in our project is to refactor code (I'm doing it since some years now), so I have less work later if the code kept clean automatically (Build-
 server with Google Test, warnings as errors, CppCheck, clang-tidy and in near future clang-format).

And clang-tidy is nice, because e.g. when to use `std::move` and when const reference is sometimes hard to know, because I don't know of every class of us if it is trivially copyable or not.

**Comment by taketwo** [comment]
> performance: 1 367

Curious to see those :)

**Comment by SunBlack** [comment]
> > performance: 1 367
> 
> Curious to see those :)

With deduplication these are not this much;
* performance-implicit-conversion-in-loop: 1
* performance-inefficient-string-concatenation: 4
* performance-type-promotion-in-math-fn: 2
* performance-unnecessary-copy-initialization:8
* performance-unnecessary-value-param: 359

**Comment by gnawme** [comment]
Easiest way:

- Add `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` to `CMakeLists.txt`
- Run `cmake .`
- This generates `compile_commands.json`
- Use `run-clang-tidy`, which will look for, and iterate through, `compile_commands.json` and apply the specified checks
- Policy at my employer is to run `clang-tidy` (a subset of the `modernize-*` checks) when your branch is close to being ready to merge (running it in CI for every push is not useful)

**Comment by kunaltyagi** [comment]
Is there a way to generate `compile_commands.json` without requiring all dependencies to be installed?

**Comment by gnawme** [comment]
> Is there a way to generate `compile_commands.json` without requiring all dependencies to be installed?

`set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` just generates the compile database, I'm not sure that it's analyzing or requiring dependencies as well.

**Comment by SunBlack** [comment]
> Is there a way to generate `compile_commands.json` without requiring all dependencies to be installed?

Yes, as it requires a successful run of CMake. Furthermore clang-tidy requires also all dependencies, to can check the code (not like CppCheck).

**Comment by gnawme** [comment]
> > Is there a way to generate `compile_commands.json` without requiring all dependencies to be installed?
> 
> Yes, as it requires a successful run of CMake. Furthermore clang-tidy requires also all dependencies, to can check the code (not like CppCheck).

Generating `compile_commands.json` only requires that `cmake .` complete, is that what you mean? It will often complain that certain dependencies are missing, but still generate the compile database.

**Comment by SunBlack** [comment]
> Generating `compile_commands.json` only requires that `cmake .` complete, is that what you mean? It will often complain that certain dependencies are missing, but still generate the compile database.

Yes, but when a dependency is missing, the target which requires these are not part of the database, as we skip some CMake code than.

**Comment by gnawme** [comment]
Opened PR #4560, working from the [base modules of the PCL dependency graph](https://user-images.githubusercontent.com/1709142/87282880-fd5cc080-c4f4-11ea-878e-9c7de9c71eb8.png) to keep the LOC changed to a manageable quantity. Future PRs will continue up the dependency graph.
